File: matlab/code/8-removing-noise.m
clc;
clear;
close all;

% Sampling frequency
Fs = 8000;

% Duration
duration = 2;

% Time vector
t = 0:1/Fs:duration-1/Fs;

% Generate 440 Hz audio signal
f0 = 440;
clean_signal = sin(2*pi*f0*t);

% Generate random noise
noise = 0.5 * randn(size(t));

% Noisy signal
noisy_signal = clean_signal + noise;

% DFT
N = length(noisy_signal);
X = fft(noisy_signal);

% Frequency vector
f = (0:N-1) * Fs/N;

% High frequency filtering
cutoff_frequency = 1000;

X_filtered = X;

% Remove frequencies above cutoff
X_filtered(f > cutoff_frequency) = 0;

% Also remove corresponding negative frequencies
X_filtered(f > Fs-cutoff_frequency) = 0;

% IDFT
cleaned_signal = real(ifft(X_filtered));

% Plot clean signal
figure;
plot(t, clean_signal);
grid on;
xlabel('Time (s)');
ylabel('Amplitude');
title('Original 440 Hz Signal');
xlim([0 0.02]);

% Plot noisy signal
figure;
plot(t, noisy_signal);
grid on;
xlabel('Time (s)');
ylabel('Amplitude');
title('Noisy Signal');
xlim([0 0.02]);

% Plot cleaned signal
figure;
plot(t, cleaned_signal);
grid on;
xlabel('Time (s)');
ylabel('Amplitude');
title('Cleaned Signal');
xlim([0 0.02]);

% Frequency spectrum
figure;
plot(f(1:floor(N/2)), abs(X(1:floor(N/2)))/N);
grid on;
xlabel('Frequency (Hz)');
ylabel('Magnitude');
title('Frequency Spectrum of Noisy Signal');
xlim([0 2000]);

File: python/code/8-removing-noise.py
import numpy as np
import matplotlib.pyplot as plt

# Sampling frequency
fs = 1000

# Duration
duration = 1

# Time
t = np.arange(0, duration, 1 / fs)

# -------------------------
# 1. Generate 440 Hz signal
# -------------------------
f = 440

pure_signal = np.sin(2 * np.pi * f * t)


# -------------------------
# 2. Add random noise
# -------------------------
np.random.seed(10)

noise = 0.5 * np.random.randn(len(t))

noisy_signal = pure_signal + noise


# -------------------------
# 3. DFT
# -------------------------
N = len(noisy_signal)

X = np.fft.fft(noisy_signal)

frequencies = np.fft.fftfreq(N, 1 / fs)


# -------------------------
# 4. High-frequency filtering
# -------------------------
cutoff = 450

X_filtered = X.copy()

X_filtered[np.abs(frequencies) > cutoff] = 0


# -------------------------
# 5. IDFT
# -------------------------
filtered_signal = np.fft.ifft(X_filtered)

filtered_signal = np.real(filtered_signal)


# -------------------------
# Plot
# -------------------------
plt.figure(figsize=(12, 10))

plt.subplot(3, 1, 1)
plt.plot(t, pure_signal)
plt.title("Pure 440 Hz Sine Wave")
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")
plt.grid(True)

plt.subplot(3, 1, 2)
plt.plot(t, noisy_signal)
plt.title("Noisy Signal")
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")
plt.grid(True)

plt.subplot(3, 1, 3)
plt.plot(t, filtered_signal)
plt.title("Filtered Signal")
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")
plt.grid(True)

plt.tight_layout()
plt.show()